engines/config-query-sparql/config/rdf-resolve-quad-pattern/actors.json

```diff
@@ -1,10 +1,43 @@
 {
   "@context": [
-    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/config-query-sparql/^2.0.0/components/context.jsonld"
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/runner/^2.0.0/components/context.jsonld",
+
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/config-query-sparql/^2.0.0/components/context.jsonld",
+
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-rdf-resolve-quad-pattern-string-source/^2.0.0/components/context.jsonld",
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-rdf-resolve-quad-pattern-federated/^2.0.0/components/context.jsonld",
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-rdf-resolve-quad-pattern-hypermedia/^2.0.0/components/context.jsonld",
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-rdf-resolve-quad-pattern-rdfjs-source/^2.0.0/components/context.jsonld"
   ],
-  "import": [
```

Comment:
I was not able to define the beforeActor otherwise, I'm gonna try it again.

Reply:
I was not able to find an example with `beforeActor` in a separate config file and it doesn't seems to be working. I'm getting
```
Detected fatal error. Generated 'componentsjs-error-state.json' with more information.
ErrorResourcesContext: The value "[urn:comunica:default:rdf-resolve-quad-pattern/actors#hypermedia]" for parameter "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-rdf-resolve-quad-pattern-rdfjs-source/^2.0.0/components/ActorRdfResolveQuadPatternRdfJsSource.jsonld#ActorRdfResolveQuadPatternRdfJsSource_args_beforeActors" is not of required range type "(https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor)<UNKNOWN GENERIC: https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-abstract-mediatyped/^2.0.0/components/ActorAbstractMediaTyped.jsonld#ActorAbstractMediaTyped__generic_I, UNKNOWN GENERIC: https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-abstract-mediatyped/^2.0.0/components/ActorAbstractMediaTyped.jsonld#ActorAbstractMediaTyped__generic_T, UNKNOWN GENERIC: https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-abstract-mediatyped/^2.0.0/components/ActorAbstractMediaTyped.jsonld#ActorAbstractMediaTyped__generic_O>[] | undefined"
    at Function.throwIncorrectTypeError (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/parameterproperty/ParameterPropertyHandlerRange.js:399:15)
    at ParameterPropertyHandlerRange.captureType (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/parameterproperty/ParameterPropertyHandlerRange.js:37:39)
    at ParameterPropertyHandlerRange.handle (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/parameterproperty/ParameterPropertyHandlerRange.js:19:14)
    at ParameterHandler.applyParameterValues (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/ParameterHandler.js:52:33)
    at ConfigPreprocessorComponentMapped.getParameterValue (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/ConfigPreprocessorComponentMapped.js:56:18)
    at ConstructorArgumentsElementMappingHandlerKeyValue.handleKeyValue (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/constructorargumentsmapping/ConstructorArgumentsElementMappingHandlerKeyValue.js:30:30)
    at ConstructorArgumentsElementMappingHandlerKeyValue.handle (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/constructorargumentsmapping/ConstructorArgumentsElementMappingHandlerKeyValue.js:24:25)
    at ConfigPreprocessorComponentMapped.applyConstructorArgumentsParameters (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/ConfigPreprocessorComponentMapped.js:43:32)
    at ConstructorArgumentsElementMappingHandlerFields.handle (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/constructorargumentsmapping/ConstructorArgumentsElementMappingHandlerFields.js:24:18)
    at ConfigPreprocessorComponentMapped.applyConstructorArgumentsParameters (/home/idlab357/Documents/PhD/comunica/node_modules/componentsjs/lib/preprocess/ConfigPreprocessorComponentMapped.js:43:32)
```

Reply:
the error file

```
{
  "cause": {
    "description": "no union values are valid",
    "causes": [
      {
        "description": "one or more array values are invalid",
        "causes": [
          {
            "description": "generic component is invalid",
            "causes": [
              {
                "description": "value is not a subtype of \"https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor\"",
                "context": {
                  "value": {
                    "@id": "urn:comunica:default:rdf-resolve-quad-pattern/actors#hypermedia",
                    "properties": {
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#genericTypeInstancesComponentScope": [
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor"
                        }
                      ],
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#genericTypeInstance": [
                        {
                          "@id": "_:df_3_2255"
                        },
                        {
                          "@id": "_:df_3_2256"
                        },
                        {
                          "@id": "_:df_3_2257"
                        }
                      ]
                    }
                  },
                  "type": {
                    "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor",
                    "properties": {
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#componentPath": [
                        "\"Actor\""
                      ],
                      "http://www.w3.org/2000/01/rdf-schema#subClassOf": [
                        {
                          "@id": "_:df_691_3"
                        }
                      ],
                      "http://www.w3.org/2000/01/rdf-schema#comment": [
                        "\"An actor can act on messages of certain types and provide output of a certain type.  The flow of an actor is as follows: 1. Send a message to {@link Actor#test} to test if an actor can run that action. 2. If the actor can reply to the message, let the actor run the action using {@link Actor#run}.  An actor is typically subscribed to a bus, using which the applicability to an action can be tested.\""
                      ],
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#genericTypeParameter": [
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__generic_I"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__generic_T"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__generic_O"
                        }
                      ],
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#parameter": [
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor_args_name"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor_args_bus"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor_args_beforeActors"
                        }
                      ],
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#memberField": [
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_name"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_bus"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_beforeActors"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_constructor"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_getContextLogger"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_test"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_run"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_runObservable"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_initialize"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_deinitialize"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_getDefaultLogData"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_logTrace"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_logDebug"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_logInfo"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_logWarn"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_logError"
                        },
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core/^2.0.0/components/Actor.jsonld#Actor__member_logFatal"
                        }
                      ],
                      "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#constructorArguments": [
                        {
                          "@id": "_:df_691_38"
                        }
                      ],
                      "http://www.w3.org/1999/02/22-rdf-syntax-ns#type": [
                        "https://linkedsoftwaredependencies.org/vocabularies/object-oriented#AbstractClass"
                      ],
                      "module": [
                        {
                          "@id": "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/core"
                        }
                      ]
                    }
                  }
                }
              }
            ]
          }
        ]
      }
    ]
  },
```

Reply:
Could you push your changes so I can see how you tried it?

Reply:
> Could you push your changes so I can see how you tried it?

I've created another branch instead

https://github.com/constraintAutomaton/comunica/tree/experiment/beforeActor-separate-config-file

Reply:
and a pool request in case it's more convenient to see.

https://github.com/constraintAutomaton/comunica/pull/1

Reply:
Ah yes, we have exactly the same problem as before :-)
During the compilation of the query-sparql-rdfjs engine, only the rdfjs file is loaded, and not the hypermedia file, so `urn:comunica:default:rdf-resolve-quad-pattern/actors#hypermedia` will be undefined.

So instead of defining `beforeActors` in the actor file, we should place it in a separate file (which is not loaded for the query-sparql-rdfjs engine).
We can solve this by adding this info to `actors.json`.
So concretely, I would suggest keeping the import-based `actors.json` file, but do something roughly like this:
```json
{
  "@context": [
    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/config-query-sparql/^2.0.0/components/context.jsonld"
  ],
  "graph": [
    {
      "import": [
        "ccqs:config/rdf-resolve-quad-pattern/actors/federated.json",
        "ccqs:config/rdf-resolve-quad-pattern/actors/hypermedia.json",
        "ccqs:config/rdf-resolve-quad-pattern/actors/rdfjs-source.json",
        "ccqs:config/rdf-resolve-quad-pattern/actors/string-source.json"
      ]
    },
    {
      "@id": "urn:comunica:default:rdf-resolve-quad-pattern/actors#string-source",
      "@type": "ActorRdfResolveQuadPatternStringSource",
      "beforeActors": { "@id": "urn:comunica:default:rdf-resolve-quad-pattern/actors#hypermedia" }
    }
  ]
}
```
So that the beforeActors relationships are defined in this file, instead of the separate actor files.

Note that actors.json only contains references to the actors (`@id`+`@type`), while the other parameters are defined in the separate actor files.

This is probably a general pattern we should follow throughout all config files.

Reply:
> Ah yes, we have exactly the same problem as before :-) During the compilation of the query-sparql-rdfjs engine, only the rdfjs file is loaded, and not the hypermedia file, so `urn:comunica:default:rdf-resolve-quad-pattern/actors#hypermedia` will be undefined.
> 
> So instead of defining `beforeActors` in the actor file, we should place it in a separate file (which is not loaded for the query-sparql-rdfjs engine). We can solve this by adding this info to `actors.json`. So concretely, I would suggest keeping the import-based `actors.json` file, but do something roughly like this:
> 
> ```json
> {
>   "@context": [
>     "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/config-query-sparql/^2.0.0/components/context.jsonld"
>   ],
>   "graph": [
>     {
>       "import": [
>         "ccqs:config/rdf-resolve-quad-pattern/actors/federated.json",
>         "ccqs:config/rdf-resolve-quad-pattern/actors/hypermedia.json",
>         "ccqs:config/rdf-resolve-quad-pattern/actors/rdfjs-source.json",
>         "ccqs:config/rdf-resolve-quad-pattern/actors/string-source.json"
>       ]
>     },
>     {
>       "@id": "urn:comunica:default:rdf-resolve-quad-pattern/actors#string-source",
>       "@type": "ActorRdfResolveQuadPatternStringSource",
>       "beforeActors": { "@id": "urn:comunica:default:rdf-resolve-quad-pattern/actors#hypermedia" }
>     }
>   ]
> }
> ```
> 
> So that the beforeActors relationships are defined in this file, instead of the separate actor files.
> 
> Note that actors.json only contains references to the actors (`@id`+`@type`), while the other parameters are defined in the separate actor files.
> 
> This is probably a general pattern we should follow throughout all config files.

when I'm trying that I'm receiving

```
Error: Error while parsing file "/engines/config-query-sparql/config/rdf-resolve-quad-pattern/actors.json": Invalid predicate IRI: graph
```

Reply:
Ah, my bad, that needs to be `@graph` instead.

Reply:
It still doesn't work for another reason

```
Error: Error while parsing file "/home/idlab357/Documents/PhD/comunica/engines/config-query-sparql/config/rdf-resolve-quad-pattern/actors.json": Invalid predicate IRI: beforeActors
```

Reply:
You may require some other changes here and there. Perhaps also some imports.

Reply:
> You may require some other changes here and there. Perhaps also some imports.

I've solved the problem.

Reply:
🥳